cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsKey.java

```diff
@@ -70,28 +73,52 @@ public static String getMetadataKey(String type, byte[] rowKey) {
 
   public static String getTargetType(byte[] rowKey) {
     MDSKey.Splitter keySplitter = new MDSKey(rowKey).split();
-    // The rowkey is [rowPrefix][targetType][targetId][metadata-type][key]
+    // The rowkey is [rowPrefix][targetType][targetId][key] for value rows and
+    // [rowPrefix][targetType][targetId][key][index] for value index rows
     keySplitter.getBytes();
     return keySplitter.getString();
   }
 
-  public static MDSKey getMDSKey(Id.NamespacedId targetId, @Nullable String key) {
-    String targetType = KeyHelper.getTargetType(targetId);
-    MDSKey.Builder builder = new MDSKey.Builder();
-    builder.add(ROW_PREFIX);
-    builder.add(targetType);
-    KeyHelper.addNamespaceIdToKey(builder, targetId);
+  /**
+   * Creates a key for metadata value row in the format:
+   * [{@link #VALUE_ROW_PREFIX}][targetType][targetId][key] for value index rowss
+   */
+  public static MDSKey getMDSValueKey(Id.NamespacedId targetId, @Nullable String key) {
+    MDSKey.Builder builder = getMDSKeyPrefix(targetId, VALUE_ROW_PREFIX);
     if (key != null) {
       builder.add(key);
     }
+    return builder.build();
+  }
 
+  /**
+   * Creates a key for metadata index row in the format:
+   * [{@link #INDEX_ROW_PREFIX}][targetType][targetId][key][index] for value index rows
+   */
+  public static MDSKey getMDSIndexKey(Id.NamespacedId targetId, String key, @Nullable String index) {
+    MDSKey.Builder builder = getMDSKeyPrefix(targetId, INDEX_ROW_PREFIX);
+    builder.add(key);
+    if (index != null) {
+      builder.add(index);
+    }
     return builder.build();
   }
 
+  private static MDSKey.Builder getMDSKeyPrefix(Id.NamespacedId targetId, byte[] indexRowPrefix) {
+    String targetType = KeyHelper.getTargetType(targetId);
+    MDSKey.Builder builder = new MDSKey.Builder();
+    builder.add(indexRowPrefix);
+    builder.add(targetType);
+    KeyHelper.addNamespaceIdToKey(builder, targetId);
```

Comment:
not your PR, but this method name is confusing, since it seems to indicate that it adds a namespace id to the key, when it actually adds an ``Id.NamespacedId``. Better to call it ``addEntityIdToKey``?